General coding practices/Test Automation (Mobile) - element IDs.md

```diff
@@ -0,0 +1,98 @@
+
+While implementing IDs, the developer should sync with the responsible tester who will be doing the automated tests for IDs and implementation limitations. There shouldn’t be any compromises to the quality of features and usability when implementing these IDs.
+
+### Naming
+
+The IDs added to elements should be **unique** (most importantly - be unique on a specific screen). For that purpose, **prefixes and/or suffixes** can be used to define the ID of the component more precisely. That will be especially helpful in the case when the component consists of other sub-components which will also have IDs added to them.
+
+*Example:* 
+
+<span style="display:block; border: 1px solid #e0e0e0; margin-top:15px; margin-bottom:15px; margin-left:auto; margin-right:auto; width:80%;">![Android Naming Example](/img/test_automation/TA_Naming.png)</span>
+
+Developers can follow naming conventions for their respective platform (*camelCase*, *snake_case*, or other).
+
+## iOS
+
+### Accessibility Identifier
+
+For UI testing purposes Apple provides us with a so-called [accessibilityIdentifier](https://developer.apple.com/documentation/uikit/uiaccessibilityidentification/1623132-accessibilityidentifier). It is meant to "be used to uniquely identify a UI element in the scripts we write using the UI Automation interfaces". Accessibility identifiers aren't accessed by e.g. VoiceOver. They are part of the `UIAccessibilityIdentification` protocol, which consists of methods that associate a **unique identifier** with elements in a user interface. The only strict requirement for adhering to this protocol is defining the `accessibilityIdentifier` property.
+
+`UIAccessibilityIdentification: var accessibilityIdentifier: String? { get set }`
+
+The mobile developer needs to set accessibilityIdentifiers on elements that are being inspected as a part of test automation (e.g., buttons, input fields, list containers and items, titles, etc.). An accessibility identifier can be set using storyboards or programmatically. In our app, we should set it programmatically to have more readable code and to make changes easier if some identifier needs to be updated.
+
+For list items that are not known in advance, the developer should implement an ID to the list container. After that, the testers should be able to get those elements as children of the container. Also, in case of adding IDs to individual list items that are not known in advance, it is alright if the same ID is implemented to them because the automation framework can locate such elements by combining the ID and text of the element.
```

Comment:
Maybe it's just me, but the `the developer` and `the tester` narrative throughout the chapter is a bit weird. It seems like we are doing some kind of a division between those two folks on the project. For example this one:

```
For list items that are not known in advance, the developer should implement an ID to the list container.
```

I would rework it a bit to

```
For list items that are not known in advance, developers should implement an ID into the list container.
```
 or
```
For list items that are not known in advance, ID should be added to te list container.
```

What do you think about it? I do not insist on this one!

Reply:
Completely agree with you. 😄 
I updated it in the whole text and left "developer/tester" only in the cases where it is needed for semantic reasons. The rest is updated (either passive or generic forms are used). Feel free to check it out.